lib/src/main/java/com/cloudbees/groovy/cps/impl/BlockScopeEnv.java

```diff
@@ -12,16 +13,34 @@
  */
 // TODO: should be package local once all the impls move into this class
 public class BlockScopeEnv extends ProxyEnv {
-    private final Map<String,Object> locals = new HashMap<String, Object>();
-    private Map<String, Class> types = new HashMap<String, Class>();
+    private Map<String,Object> locals;
+    private Map<String, Class> types;
 
     public BlockScopeEnv(Env parent) {
+        this(parent, 0);
+    }
+
+    public BlockScopeEnv(Env parent, int localsSize) {
         super(parent);
+        if (localsSize <= 0) {
+            locals = Collections.emptyMap();
+            types = Collections.emptyMap();
+        } else {
+            locals = new HashMap<String, Object>(localsSize);
```

Comment:
I think you are not understanding `initialCapacity`. If you intended `localsSize` to be the expected number of entries, and you want to avoid rehashing, then you will need to use something like `localsSize * 4 / 3 + 1`.